android/src/com/frostwire/android/gui/fragments/preference/ApplicationFragment.java

```diff
@@ -67,9 +71,40 @@ public ApplicationFragment() {
     protected void initComponents() {
         setupConnectSwitch();
         setupStorageOption();
+        setupDataSaving();
         setupStore(removeAdsPurchaseTime);
     }
 
+    private void setupDataSaving() {
+        SwitchPreferenceCompat preference = findPreference(Constants.PREF_KEY_NETWORK_USE_WIFI_ONLY);
+        preference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
+            @Override
+            public boolean onPreferenceChange(final Preference preference, Object newValue) {
+                boolean newVal = (Boolean) newValue;
+                if (newVal && !NetworkManager.instance().isDataWIFIUp()) {
+                    if (TransferManager.instance().isHttpDownloadInProgress()) {
+                        UIUtils.showYesNoDialog(getActivity(), R.string.data_saving_kill_http_warning, R.string.data_saving_kill_http_warning_title, new DialogInterface.OnClickListener() {
+                            @Override
+                            public void onClick(DialogInterface dialog, int which) {
+                                turnOffTransfers();
+                                setChecked((TwoStatePreference) preference, true, false);
+                            }
+                        });
+                        return false;
+                    }
+                    turnOffTransfers();
+                }
+                return true;
+            }
+        });
+    }
+
+    private void turnOffTransfers() {
+        TransferManager.instance().stopHttpTransfers();
+        TransferManager.instance().pauseTorrents();
```

Comment:
those packets will not travel if not in Wi-Fi, not clear what happens in network reconnection